Mathlib/Topology/GPseudoMetric/Basic.lean

```diff
@@ -0,0 +1,332 @@
+/-
+Copyright (c) 2024 Edward van de Meent. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Edward van de Meent
+-/
+import Mathlib.Algebra.CovariantAndContravariant
+import Mathlib.Algebra.NeZero
+import Mathlib.Algebra.Order.Monoid.Lemmas
+import Mathlib.Data.Set.Lattice
+
+/-!
+# Pseudo-metrics with generic codomain
+
+In this file we introduce `GPseudoMetric`.
+This differs from `PseudoMetricSpace` by not requiring that the codomain of the metric be `ℝ`,
+instead only requiring that it is an (additive) commutative monoid, with a linear ordering.
+
+## Main definitions
+
+- `GPseudoMetric α β`: a structure containing a distance function on `α` with codomain `β`,
+which may be equal to `0` for non-equal elements. The distance function is `0` for equal elements,
+is commutative in its arguments, and satisfies the triangle inequality.
+- `GPseudoMetricClass α β`: the class of types of generic pseudo-metrics on `α` to `β`.
+
+Additional useful definitions:
+
+- `ball gdist x δ`: the set of points with distance to `x` strictly less than `δ`
+- `closedBall gdist x δ`: the set of points with distance to `x` less than or equal to `δ`
+- `sphere gdist x δ`: the set of points with distance to `x` equal to `δ`
```

Comment:
Please fully qualify names so that the links work in the docs
```suggestion
- `GMetric.ball gdist x δ`: the set of points with distance to `x` strictly less than `δ`
- `GMetric.closedBall gdist x δ`: the set of points with distance to `x` less than or equal to `δ`
- `GMetric.sphere gdist x δ`: the set of points with distance to `x` equal to `δ`
```